libafl/src/events/broker_hooks/centralized_multi_machine.rs

```diff
@@ -0,0 +1,281 @@
+#[derive(Debug)]
+pub struct TcpMultiMachineLlmpReceiverHook<A, I>
+where
+    I: Input,
+{
+    /// the actual state of the broker hook
+    shared_state: Arc<RwLock<TcpMultiMachineState<A>>>,
+    /// the tokio runtime used to interact with other machines. Keep it outside to avoid locking it.
+    rt: Arc<Runtime>,
+    phantom: PhantomData<I>,
+}
+
+impl<A, I> TcpMultiMachineLlmpSenderHook<A, I>
+where
+    A: Clone + Display + ToSocketAddrs + Send + Sync + 'static,
+    I: Input + Send + Sync + 'static,
+{
+    /// Should not be created alone. Use [`TcpMultiMachineBuilder`] instead.
+    pub(crate) fn new(
+        shared_state: Arc<RwLock<TcpMultiMachineState<A>>>,
+        rt: Arc<Runtime>,
+    ) -> Self {
+        Self {
+            shared_state,
+            rt,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<A, I> TcpMultiMachineLlmpReceiverHook<A, I>
+where
+    A: Clone + Display + ToSocketAddrs + Send + Sync + 'static,
+    I: Input + Send + Sync + 'static,
+{
+    /// Should not be created alone. Use [`TcpMultiMachineBuilder`] instead.
+    pub(crate) fn new(
+        shared_state: Arc<RwLock<TcpMultiMachineState<A>>>,
+        rt: Arc<Runtime>,
+    ) -> Self {
+        Self {
+            shared_state,
+            rt,
+            phantom: PhantomData,
+        }
+    }
+
+    #[cfg(feature = "llmp_compression")]
+    fn try_compress(
+        state_lock: &mut RwLockWriteGuard<TcpMultiMachineState<A>>,
+        event: &Event<I>,
+    ) -> Result<(Flags, Vec<u8>), Error> {
+        let serialized = postcard::to_allocvec(&event)?;
+
+        match state_lock.compressor().maybe_compress(&serialized) {
+            Some(comp_buf) => Ok((LLMP_FLAG_COMPRESSED, comp_buf)),
+            None => Ok((Flags(0), serialized)),
+        }
+    }
+
+    #[cfg(not(feature = "llmp_compression"))]
+    fn try_compress(
+        _state_lock: &mut RwLockWriteGuard<TcpMultiMachineState<A>>,
+        event: &Event<I>,
+    ) -> Result<(Flags, Vec<u8>), Error> {
+        Ok((Flags(0), postcard::to_allocvec(&event)?))
+    }
+}
+
+impl<A, I, SP> LlmpHook<SP> for TcpMultiMachineLlmpSenderHook<A, I>
+where
+    A: Clone + Debug + Display + ToSocketAddrs + Send + Sync + 'static,
+    SP: ShMemProvider,
+    I: Input + Send + Sync + 'static,
+{
+    /// check for received messages, and forward them alongside the incoming message to inner.
+    fn on_new_message(
+        &mut self,
+        _broker_inner: &mut LlmpBrokerInner<SP>,
+        _client_id: ClientId,
+        _msg_tag: &mut Tag,
+        _msg_flags: &mut Flags,
+        msg: &mut [u8],
+        _new_msgs: &mut Vec<(Tag, Flags, Vec<u8>)>,
+    ) -> Result<LlmpMsgHookResult, Error> {
+        let shared_state = self.shared_state.clone();
+
+        // Here, we suppose msg will never be written again and will always be available.
+        // Thus, it is safe to handle this in a separate thread.
+        let msg_lock = unsafe { NullLock::new((msg.as_ptr(), msg.len())) };
```

Comment:
I feel like this is wrong on a rust level, the mutable borrow only lives as long as `on_new_message`, right? And here you spawn an async function that runs _after_ the function returns, using the `msg`.

_at least_ the `TcpMultiMachineLlmpReceiverHook` new function should be `unsafe`, and this side effect ("on_new_message needs msg to have a quasi-static lifetime) should be documented there

Reply:
Please fix

Reply:
the msg_lock will leave beyond the call to on_new_message. it's valid with 2 hypothesis:
- messages, once sent, are never written again
- once a message is posted, it is never deallocated.

this patch is very experimental in general, we merged for practical reasons (it's part of the reason why it's hidden behind an undocumented feature for now)

I'll open the follow-up patch